File: 54_subsets.js
/*
Given an integer array nums of unique elements, return all possible subsets
(the power set).

The solution set must not contain duplicate subsets. 
Return the solution in any order.
*/

// https://leetcode.com/problems/subsets/description/

// backtracking without duplicates

var subsets = function (nums) {
  let path = [],
    result = [];
  nums.sort((a, b) => a - b);
  var dfs = function (nums, start) {
    result.push([...path]);
    for (let i = start; i < nums.length; i++) {
      path.push(nums[i]);
      dfs(nums, i + 1);
      path.pop();
    }
  };
  dfs(nums, 0);
  return result;
};

File: 54_subsets.py
# Given an integer array nums of unique elements, return all possible subsets
# (the power set).

# The solution set must not contain duplicate subsets. 
# Return the solution in any order.

# https://leetcode.com/problems/subsets/
# advanced LC 90 https://leetcode.com/problems/subsets-ii/

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res, path, l = [], [], len(nums)
        def backtracking(idx):
            # base case
            res.append(path[:])
            if len(path) == l: return
            for i in range(idx, l):
                path.append(nums[i])
                backtracking(i + 1)
                path.pop()
        
        backtracking(0)
        return res